internal/locate/region_cache.go

```diff
@@ -1240,25 +1240,30 @@ func (c *RegionCache) BatchLocateKeyRanges(bo *retry.Backoffer, keyRanges []kv.K
 				keyRange.StartKey = lastRegion.EndKey()
 			}
 		}
-		for {
-			// TODO: find all the cached regions in the range.
-			// now we only check if the region is cached from the lower bound, if there is a uncached hole in the middle,
-			// we will load the rest regions even they are cached.
-			r := c.tryFindRegionByKey(keyRange.StartKey, false)
-			lastRegion = r
-			if r == nil {
-				// region cache miss, add the cut range to uncachedRanges, load from PD later.
-				uncachedRanges = append(uncachedRanges, pd.KeyRange{StartKey: keyRange.StartKey, EndKey: keyRange.EndKey})
+		// TODO: find all the cached regions in the range.
+		// now we only check if the region is cached from the lower bound, if there is a uncached hole in the middle,
+		// we will load the rest regions even they are cached.
+		batchRegionInCache, err := c.scanRegionsFromCache(bo, keyRange.StartKey, keyRange.EndKey)
+		if err != nil {
+			return nil, err
+		}
+		hit := false
+		for _, r := range batchRegionInCache {
+			if !r.Contains(keyRange.StartKey) { // uncached hole, load the rest regions
 				break
 			}
-			// region cache hit, add the region to cachedRegions.
 			cachedRegions = append(cachedRegions, r)
+			lastRegion = r
 			if r.ContainsByEnd(keyRange.EndKey) {
 				// the range is fully hit in the region cache.
+				hit = true
```

Comment:
maybe rename `hit` to `containsAll` can be more understandable?

Reply:
Addressed.